File: tree/python/leetcode101_Symmetric_Tree.py

# Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

# For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

#     1
#    / \
#   2   2
#  / \ / \
# 3  4 4  3
# But the following [1,2,2,null,3,null,3] is not:
#     1
#    / \
#   2   2
#    \   \
#    3    3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        stack = [root]
        while stack:
            for i in range(len(stack)//2):
                if stack[i] and stack[~i]:
                    if stack[i].val != stack[~i].val:
                        return False
                elif stack[i] or stack[~i]:
                    return False
            newstack = []
            for node in stack:
                if node:
                    newstack += [node.left, node.right]
            stack = newstack
        return True



    def isSymmetric1(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root: return True
        return self.helper(root.left, root.right)
    
    def helper(self, p, q):
        if not p and not q:
            return True
        if not p or not q:
            return False
        return q.val == p.val and self.helper(p.left, q.right) and self.helper(p.right, q.left)

# Time: O(n)
# Space: O(n)
# Difficulty: easy

File: tree/java/leetcode101_Symmetric_Tree.java
//
//Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).
//
//For example, this binary tree [1,2,2,3,4,4,3] is symmetric:
//
//    1
//   / \
//  2   2
// / \ / \
//3  4 4  3
//But the following [1,2,2,null,3,null,3] is not:
//    1
//   / \
//  2   2
//   \   \
//   3    3

package wenxinjie1;

import java.util.LinkedList;
import java.util.Queue;

public class leetcode101_Symmetric_Tree {
    public boolean isSymmetric(TreeNode root) {
        if(root==null) return true;
        return isMirror(root.left,root.right);
}
    public boolean isMirror(TreeNode p, TreeNode q) {
        if(p==null && q==null) return true;
        if(p==null || q==null) return false;
        return (p.val==q.val) && isMirror(p.left,q.right) && isMirror(p.right,q.left);
}
    
    public boolean isSymmetric1(TreeNode root) { 	
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        if(root == null) return true;
        q.add(root.left);
        q.add(root.right);
        while (q.size()>1){
            TreeNode left = q.poll(), right = q.poll();
            if (left == null && right == null) continue;
            if (left == null || right == null) return false;
            if (left.val != right.val) return false;
            q.add(left.left);
            q.add(right.right);
            q.add(left.right);
            q.add(right.left);           
        }
        return true;
    }
}

//Time: O(n)
//Space: O(n)
//Difficulty: easy